ecl/hql/hqlexpr.cpp

```diff
@@ -16188,6 +16618,292 @@ IHqlExpression * annotateIndexBlobs(IHqlExpression * expr)
+                            d[im1][j  ] + 1,
+                            d[im1][jm1] + cost);
+            }
+            d[i][j] = next;
+
+            if (i1 > 0 && j1 > 0)
+            {
+                unsigned char transposition = d[i1-1][j1-1] + (im1-i1) + (jm1-j1) + cost;
+                if (d[i][j] > transposition)
+                    d[i][j] = transposition;
+            }
+        }
+        c[left[im1]] = i;
+    }
+
+    return d[leftLen][rightLen];
+}
+
+void NearestSymbol::compare(NearestSymbol & comparison)
+{
+    unsigned penalty = 0;
+    compare(comparison, penalty);
+}
+
+void NearestSymbol::compare(NearestSymbol & comparison, unsigned & penalty)
+{
+    if (!comparison.nearest)
+        return;
+
+    if (nearest == comparison.nearest)
+        return;
+
+    unsigned unsignedDistance = comparison.distance == (unsigned)(-1) ? comparison.distance : comparison.distance + penalty;
+    if (unsignedDistance < distance)
+    {
+        distance = comparison.distance;
+        nearest.set(comparison.nearest);
+        penalty++;
+    }
+}
+
+void NearestSymbol::compare(IHqlExpression * comparison)
+{
+    if (!comparison || nearest == comparison)
+        return;
+
+    unsigned tempDistance = computeEditDistance(comparison->queryId());
+
+    if (tempDistance == 0 || tempDistance >= MAX_SYMBOL_DISTANCE)
+        return;
+
+    if (tempDistance < distance)
+    {
+        distance = tempDistance;
+        nearest.set(comparison);
+    }
+}
+
+unsigned NearestSymbol::computeEditDistance(const IIdAtom * comparison) const
+{
+    if (!comparison)
+        return -1;
+
+    const char * comparisonLowerStr = str(lower(comparison));
+    const char * searchNameLowerStr = str(lower(searchName));
+    if (strcmp(searchNameLowerStr, comparisonLowerStr) == 0)
+        return 0;
+
+    //The edit distance computation below is the full Damerau Levenshtein Distance with a weighted matrix.
+    //Therefore it is possible for it to return a distance of 0 yet the two strings not be identicle.
+    //For this reason +1 is added to differentiate if no weights are used, the restricted version is used, or the standard Levenshtein edit idstance is used.
+    switch(metric)
+    {
+    case optimalStringAlignment :
+        //call optimalStringAlignment
+    case damerauLevenshtein :
+        return damerauLevenshteinDistance(strlen(searchNameLowerStr), searchNameLowerStr, strlen(comparisonLowerStr), comparisonLowerStr, NULL, 0);
+    case weightedDamerauLevenshtein :
+        return damerauLevenshteinDistance(strlen(searchNameLowerStr), searchNameLowerStr, strlen(comparisonLowerStr), comparisonLowerStr, defaultConfusionMatrix, defaultConfusionMatrixLength) + 1;
+    };
+    return -1;
+}
+
+IHqlExpression * NearestSymbol::branchSearch(IHqlScope * scope, unsigned lookupFlags, HqlLookupContext & ctx, unsigned & penalty)
+{
+    NearestSymbol tempNearest(searchName);
+    scope->lookupNearestSymbol(searchName, lookupFlags, ctx, tempNearest);
```

Comment:
This has to be an odd github bug, as this is not the actual code, which is complete and does compile.